ext4/gpt/gpt.go

```diff
@@ -0,0 +1,269 @@
+func GetSizeOfEntryArray(numEntries int) int {
+	sizeOfEntryArrayBytes := SizeOfPartitionEntry * numEntries
+	entryArrayReservedSize := ReservedLBAsForParitionEntryArray * BlockSizeLogical
+	if sizeOfEntryArrayBytes < entryArrayReservedSize {
+		sizeOfEntryArrayBytes = entryArrayReservedSize
+	}
+	return sizeOfEntryArrayBytes
+}
+
+func GetFirstUseableLBA(sizeOfEntryArrayBytes int) uint64 {
+	totalGPTInfoSizeInBytes := sizeOfEntryArrayBytes + SizeOfHeaderInBytes
+	totaMetadataSizeInBytes := totalGPTInfoSizeInBytes + SizeOfPMBRInBytes
+
+	firstUseableLBA := FindNextUnusedLogicalBlock(uint64(totaMetadataSizeInBytes))
+	if firstUseableLBA < MinFirstUsableLBA {
+		firstUseableLBA = MinFirstUsableLBA
+	}
+	return firstUseableLBA
+}
+
+func FindNextUnusedLogicalBlock(bytePosition uint64) uint64 {
+	block := uint64(bytePosition / BlockSizeLogical)
+	if bytePosition%BlockSizeLogical == 0 {
+		return block
+	}
+
+	return block + 1
+}
+
+func CalculateEndingCHS(sizeInLBA uint32) [3]byte {
+	// See gpt package for more information on the max value of the ending CHS
+	result := [3]byte{}
+	if sizeInLBA >= ProtectiveMBREndingCHSMaxValue {
+		result = ProtectiveMBREndingCHSMaxArray
+	} else {
+		// Since sizeInLBA is 1 byte bigger than the CHS array capacity,
+		// use a temporary array to hold the result. At this point, we know
+		// that the value of sizeInLBA can be represented in the CHS array,
+		// so we can safely ignore the last value placed by binary package
+		// as it will always be 0x00
+		tempResult := [4]byte{}
+		binary.LittleEndian.PutUint32(tempResult[:], sizeInLBA)
+		copy(result[:], tempResult[:len(tempResult)-1])
+	}
+	return result
+}
+
+func CalculateHeaderChecksum(header Header) (uint32, error) {
+	buf := &bytes.Buffer{}
+	// do not include reserved field
+	if err := binary.Write(buf, binary.LittleEndian, header); err != nil {
+		return 0, err
+	}
+
+	checksum := crc32.ChecksumIEEE(buf.Bytes()[:HeaderSize])
+	return checksum, nil
+}
+
+func CalculateChecksumPartitionEntryArray(w io.ReadWriteSeeker, entryArrayLBA uint32, readLengthInBytes uint32) (uint32, error) {
+	currentBytePosition, err := w.Seek(0, io.SeekCurrent)
+	if err != nil {
+		return 0, err
+	}
+
+	// seek to position of entry array
+	entryArrayOffsetInBytes := int64(entryArrayLBA * compactext4.BlockSizeLogical)
+	_, err = w.Seek(entryArrayOffsetInBytes, io.SeekStart)
+	if err != nil {
+		return 0, err
+	}
+	buf := make([]byte, readLengthInBytes)
+	if err := binary.Read(w, binary.LittleEndian, buf); err != nil {
+		return 0, err
+	}
+
+	// calculate crc32 hash
+	checksum := crc32.ChecksumIEEE(buf)
+
+	if _, err := w.Seek(currentBytePosition, io.SeekStart); err != nil {
+		return 0, err
+	}
+	return checksum, nil
+}
+
+func ReadPMBR(r io.ReadSeeker) (ProtectiveMBR, error) {
+	current, err := r.Seek(0, io.SeekCurrent)
+	if err != nil {
+		return ProtectiveMBR{}, fmt.Errorf("failed to seek the current byte: %v", err)
+	}
+
+	pMBRByteLocation := 0 * compactext4.BlockSizeLogical
```

Comment:
What's the reasoning behind when we use `gpt.BlockSizeLogical` vs `compactext4.BlockSizeLogical`?